src/client/activation/downloader.ts

```
@@ -0,0 +1,128 @@
+        const localTempFilePath = await this.downloadFile();
+        try {
+            await this.verifyDownload(localTempFilePath);
+            await this.unpackArchive(context.extensionPath, localTempFilePath);
+        } finally {
+            fs.unlink(localTempFilePath, noop);
+        }
+    }
+
+    private async downloadFile(): Promise<string> {
+        const platformString = this.platformData.getPlatformDesignator();
+        const remoteFileName = `${downloadBaseFileName}-${platformString}.${downloadVersion}${downloadFileExtension}`;
+        const uri = `${downloadUriPrefix}/${remoteFileName}`;
+        this.output.append(`Downloading ${uri}... `);
+        const tempFile = await createTemporaryFile(downloadFileExtension);
+
+        const deferred = createDeferred();
+        const fileStream = fs.createWriteStream(tempFile.filePath);
+        fileStream.on('finish', () => {
+            fileStream.close();
+            deferred.resolve();
+        }).on('error', (err) => {
+            tempFile.cleanupCallback();
+            this.handleError(`Unable to download Python Analysis Engine. Error ${err}`);
+        });
+
+        let firstResponse = true;
+        https.get(uri, (response) => {
+            this.checkHttpResponse(response);
+            if (firstResponse) {
+                this.reportDownloadSize(response);
+                firstResponse = false;
+            }
+            response.pipe(fileStream);
+        });
+
+        await deferred.promise;
+        this.output.append('complete.');
+        return tempFile.filePath;
+    }
+
+    private async verifyDownload(filePath: string): Promise<void> {
+        this.output.appendLine('');
+        this.output.append('Verifying download... ');
+        const verifier = new HashVerifier();
+        if (!await verifier.verifyHash(filePath, this.platformData.getExpectedHash())) {
+            this.handleError('Hash of the downloaded file does not match.');

```

better to just throw an exception here, let calling code log the error.

---

It is more for the output formatting. It starts with `Verifying download... ` and I'd like it to look as 
`Verifying download... valid` on the same line and on different lines if there is an error. Same as in other cases: download and unpack.